### PR TITLE
Fix XCT fails build in F# projects

### LINKED ISSUE #1124: [Bug] Cannot build F# project after installing xamarin.communitytoolkit NuGet package
### Description
After I add xamarin.communitytoolkit package I can no longer build F# project due to .cs file.

Build error:
```
The file extension of '/Users/twop/.nuget/packages/xamarin.communitytoolkit/1.1.0/build/PreserveXamarinCommunityToolkit.cs' is not recognized. Source files must have extension .fs, .fsi, .fsx, .fsscript, .ml or .mli.
```

### Steps to Reproduce

1. Install xamarin.communitytoolkit NuGet package to F# project (.netstandard 2.0)
2. Build

### Expected Behavior
I can build the project

### Actual Behavior
There is a build error

### Basic Information

- Version with issue: 1.1
- Last known good version: -
- IDE: VS fo Mac
- Platform Target Frameworks: I use Mac target but the build issue exists on .netstandard library 
- Nuget Packages:
- Affected Devices:

### Workaround
I haven't tried referencing the .dll directly but it will probably work.

### PR DESCRIPTION
**Pull request by maxkoshevoi**
### Description of Change ###

Include `PreserveXamarinCommunityToolkit.cs` only in C# projects

Tested it with C# and F# projects:
- It's still included in C# projects
- It's no longer included in F# projects (and build no longer fails)

### Bugs Fixed ###

- Fixes #1124

### API Changes ###

None

### Behavioral Changes ###

None

### PR Checklist ###

- [ ] Has tests (if omitted, state reason in description)
- [ ] Has samples (if omitted, state reason in description)
- [x] Rebased on top of main at time of PR
- [x] Changes adhere to coding standard

- [ ] Updated [documentation](https://github.com/MicrosoftDocs/xamarin-communitytoolkit)

### CODE REVIEW

**Comment by AndreiMisiukevich** [comment]
Hi @maxkoshevoi and thanks for this contribution)
Can you please explain how did you test it? Did you create nuget packages locally?

**Comment by maxkoshevoi** [comment]
@AndreiMisiukevich Turns out our CI pipelines create NuGet for each successful run! (you can find it in artifacts)

Here's how I did the testing:
1. Created F# project and included public NuGet into it.
1) Build failed as described in the issue. 

The issue is that we are publishing `.targets` in our NuGet that are including `PreserveXamarinCommunityToolkit.cs` into every project that NuGet is added to. By the way, do you know, why is this needed?

Downloaded NuGet from pipeline artifacts for this PR
2. Included new NuGet in F# project
2) Build succeeded
3. Included new NuGet in C# project
3) Build succeeded
4. Deleted `PreserveXamarinCommunityToolkit.cs` from the NuGet's `build` folder
4) F# project still built successfully, but C# project failed with "Unable to find `PreserveXamarinCommunityToolkit.cs` error".

**Comment by AndreiMisiukevich** [comment]
Perhaps we should add a new file **PreserveXamarinCommunityToolkit.fs**
```fsharp
// <auto-generated/>

open Xamarin.CommunityToolkit.Helpers

[<assembly: PreserveToolkit>]
```

And amend targets file as

```xml
<Project>
  <ItemGroup>
    <Compile Condition="'$(MSBuildProjectExtension)' == '.csproj'" 
             Include="$(MSBuildThisFileDirectory)PreserveXamarinCommunityToolkit.cs"
             Visible="false" />
    <Compile Condition="'$(MSBuildProjectExtension)' == '.fsproj'" 
             Include="$(MSBuildThisFileDirectory)PreserveXamarinCommunityToolkit.fs"
             Visible="false" />
  </ItemGroup>
</Project>
```

And csrpoj

```xml
<None Include="build\**\*.cs;build\**\*.fs;build\**\*.targets" Pack="true" PackagePath="build" />
```

**Comment by maxkoshevoi** [comment]
Good point! Added `PreserveXamarinCommunityToolkit.fs`, but I still don't get why do we need to added it into projects that depend on the NuGet.

Testing: I'll update this comment after the testing

**Comment by AndreiMisiukevich** [comment]
Need to update Xamarin.CommunityToolkit.csproj file (at the bottom)

```csproj
<None Include="build\**\*.cs;build\**\*.fs;build\**\*.targets" Pack="true" PackagePath="build" />
```

**Comment by AndreiMisiukevich** [comment]
Need to include a new file there

image

**Comment by AndreiMisiukevich** [comment]
> Good point! Added `PreserveXamarinCommunityToolkit.fs`, but I still don't get why do we need to added it into projects that depend on the NuGet.

Frankly speaking I don't know too)

Maybe @jfversluis @pictos know better

**Comment by maxkoshevoi** [comment]
> Need to update Xamarin.CommunityToolkit.csproj file

XCT builds just fine. The problem is when I try to build F# project that includes XCT NuGet.
Those files are not included in XCT assembly anyway (since they don't have `.shared.` in them).

File: src/CommunityToolkit/Xamarin.CommunityToolkit/build/PreserveXamarinCommunityToolkit.fs [hunk 1]
```diff
@@ -0,0 +1,6 @@
+// <auto-generated/>
+
+open Xamarin.CommunityToolkit.Helpers
+
+[<assembly: PreserveToolkit>]
+do ()
```
Comment (AndreiMisiukevich) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/build/PreserveXamarinCommunityToolkit.fs:6]: I guess "do" can be omitted

**Comment by AndreiMisiukevich** [comment]
I mean, we need to add .fs file to csproj for copying it to the NuGet bundle.

File: src/CommunityToolkit/Xamarin.CommunityToolkit/build/PreserveXamarinCommunityToolkit.fs (hunk 1 above)
Comment (maxkoshevoi) [reply src/CommunityToolkit/Xamarin.CommunityToolkit/build/PreserveXamarinCommunityToolkit.fs:6]: Can it? Compilation fails without it:

image

Comment (maxkoshevoi) [reply src/CommunityToolkit/Xamarin.CommunityToolkit/build/PreserveXamarinCommunityToolkit.fs:6]: Also SO answer does it with `do`:
https://stackoverflow.com/a/2269735/4312132

Comment (AndreiMisiukevich) [reply src/CommunityToolkit/Xamarin.CommunityToolkit/build/PreserveXamarinCommunityToolkit.fs:6]: Ohhh, my bad. I proposed an incorrect fs file. Let me update it

**Comment by AndreiMisiukevich** [comment]
```fsharp
// <auto-generated/>
namespace Xamarin.CommunityToolkit.Helpers

[<assembly: PreserveToolkit>]
()
```

Please go ahead with this one

**Comment by maxkoshevoi** [comment]
@AndreiMisiukevich Also we need add `PreserveXamarinCommunityToolkit.fs` at the top somehow, since file that contains entry point must be the last in F#, and currently `PreserveXamarinCommunityToolkit.fs` is being added last, so this is another issue

**Comment by AndreiMisiukevich** [comment]
@maxkoshevoi seems it's working for me. Is it working for you now as well?

**Comment by AndreiMisiukevich** [comment]
Ah, okay, it is not... I see.
Need to find how to fix it

**Comment by maxkoshevoi** [comment]
> @maxkoshevoi seems it's working for me. Is it working for you now as well?

Got the same exception (`Files in libraries or multiple-file applications must begin with a namespace or module declaration...`). After adding `module Hi` to `Program.fs`, got this one:

image

**Comment by AndreiMisiukevich** [comment]
Manually shifting this to the top of the fsproj file solves the problem

```xml
  <Import Project="..\packages\Xamarin.CommunityToolkit.1.1.0-preview1982\build\Xamarin.CommunityToolkit.targets" Condition="Exists('..\packages\Xamarin.CommunityToolkit.1.1.0-preview1982\build\Xamarin.CommunityToolkit.targets')" />
```

**Comment by maxkoshevoi** [comment]
> Manually shifting this to the top of the fsproj file solves the problem
> 
> ```
>   <Import Project="..\packages\Xamarin.CommunityToolkit.1.1.0-preview1982\build\Xamarin.CommunityToolkit.targets" Condition="Exists('..\packages\Xamarin.CommunityToolkit.1.1.0-preview1982\build\Xamarin.CommunityToolkit.targets')" />
> ```

It doesn't for me. Are you using SDK-style `fsproj` file?

**Comment by AndreiMisiukevich** [comment]
> SDK-style

Nope, it's not SDK-style in my case

**Comment by maxkoshevoi** [comment]
It worked for me too. But location for `targets` file is different (there is no `packages` folder anywhere near the project).

Here's the path where they are located for me: `C:\Users\Mi\.nuget\packages\xamarin.communitytoolkit\1.1.0-preview1982\build\Xamarin.CommunityToolkit.targets`

**Comment by AndreiMisiukevich** [comment]
Well, let's invite more people to this conversation

CC @jfversluis @pictos @jsuarezruiz @sthewissen @brminnick

**Comment by AndreiMisiukevich** [comment]
I don't know, maybe we can skip adding this attribute for F# projects for now. Let's see what others think)

**Comment by maxkoshevoi** [comment]
> maybe we can skip adding this attribute for F# projects for now

That's what I initially thought 😄
I tried to add it, faced couple of issues and decided that it's not worth it, or someone would point out, that it needs to be done (if this is the case).

Currently I'm not even sure that we need to add it to C# projects, and I completely don't understand what it's purpose is. 
So I'm totally fine not adding it to F# projects =)

**Comment by pictos** [comment]
@jfversluis @brminnick @jsuarezruiz do you know someone on MS side that can help us with this issue?

**Review by brminnick — APPROVED** [review]

**Comment by brminnick** [comment]
Confirmed this fix with the F# team internally 👍

**Comment by brminnick** [comment]
FYI - I cherry-picked this to `develop` branch as well https://github.com/xamarin/XamarinCommunityToolkit/commit/4ee78fc9741412afc8f83546e9866a7475f01e30

**Pull request merged**